### ISSUE TITLE
Add feature to support temporary Access Key & Secret Key with session token for Amazon EC2 Node Template

### ISSUE THREAD

**Issue by gufranmmu** [rancher/rancher#15962]
Rancher Amazon EC2 Node Template doesn't support Temporary AWS AccessKey & SecretKey with session token:

![image](https://user-images.githubusercontent.com/10417811/46517557-7cc95480-c878-11e8-9c0b-4e5ed1bfafb9.png)

**Comment by vincent99** [comment]
The credentials are needed again days or months later when the instance is going to be deleted, so session tokens are not really a very good fit.

**Comment by gufranmmu** [comment]
Thanks! AWS instance profile can be used to access to EC2 API. The user can get new session tokens when days or months later instances need to be deleted. https://docs.aws.amazon.com/cli/latest/reference/sts/get-session-token.html 
This would be a killer feature from the security point of view as well since not all organizations want to give out the static access and secret keys.

**Comment by bestbear** [comment]
Yes, I just ran into this limitation in your UI today.  The place I'm working at does not give out Access/Secret keys and only allows for temporary STS generated keys.    Access/Secret keys are a huge security risk for anyone in-the-know sensitive to AWS access keys

**Comment by vincent99** [comment]
We're currently adding this for EKS 2.2, and the usability of it is going to be pretty lousy because using short-term keys in a system meant to do long-term management of resources is a fundamentally poor fit.  (and the usability of STS itself is bad in general)

Session tokens require multi factor auth for the keys it gives back to be able to do anything with IAM, which EKS needs, so you can't use them at all without also having MFA enabled.  Which your account could be configured to not even allow.  Or if it is configured, it's still not required, so you can make the wrong kind of token.  And the error you get if you provide valid credentials created without MFA is indistinguishable from the credentials being wrong entirely for a request that requires MFA.

We have to potentially ask for credentials on every update/delete operation, but at least there's not much in the way of editing an EKS cluster once it's up.  You can't do bulk operations because each thing has potentially different keys.  The APIs become a pain to use because you have to update the credentials before you can do other operations.

For RKE/the EC2 node driver this ends up making no sense until other work is done in several other areas to move cloud credentials to their own type referenced by nodes & node pools.

And at the end of the day you still end up with valid credentials stored in the the management plane a good portion of the time (if you ever use the features), so what you've actually gained from all this inconvenience of MFA+AssumeRole vs a service acct with permanent keys* and only the permissions it needs is pretty questionable IMO.  

*: which we _do_ need to provide a way to efficiently rotate, that is also in progress.

**Comment by bestbear** [comment]
How about SAML federation support then for Okta instead?  That is the real target that would allow my client to support Rancher in AWS.  Thanks

**Comment by peterfraedrich** [comment]
Why not just let the server use the IAM roles associated with the server? If I'm running my rancher server in EC2 with an IAM role that has the correct permissions I should be able to create clusters without having to provide any Keys/Tokens/Secrets.

**Comment by vincent99** [comment]
You don't have to be using a server on EC2 to create clusters on it, and you probably would want some sort of control over who is then allowed to use those credentials to spin up $all_your_quota of instances.

**Comment by cjellick** [comment]
For the reasons that @vincent99 has pointed out, this feature has a good deal of complexity to make it useable. When I asked @cloudnautique how he would want to use temporary keys in a production setup, his answer was "IAM profiles." Which is a different feature from temporary keys all together, but necessary if you want the type of security in a production setup that is being asked for here.

This effectively means that we would use the the profile of the VM on which the node provisioning was occurring. But this needs unpacked much further (as @vincent99 has pointed out). For example, do you want anyone who can create a cluster (which by default is all users who have access to Rancher) to be able to provision with the set of credentials on that box.

If we limited this solely to just implementing the temporary access key, there would be a number of UX implications. UI interactions to scale up, down, or delete a node would need to re-prompt for credentials. There are also non-UI based functions that this would affect currently occur:
- Scaling or deleting nodes can be done directly from the API
- Rancher automatically attempts to replace a node if it has been deleted.

We can do "just" temporary tokens, but basically what we would be assuming/saying is that that would result in a degraded feature set around node pools and node pool provisioning. Nevertheless, I could see that serving some value in dev/test/sandbox environments where the users are not really expecting production level node management.

**Comment by ankur111** [comment]
From the customer perspective -  sharing of permanent keys is a security issue. 
If however, we make node templates shareable and provide basic security around secrets(encryption/auto-refresh/vault integration), will it solve the problem?

**Comment by Wykiki** [comment]
Sorry for reviving an old issue, but if we add an `access_token` field, the main problem is that if we want to re-configure the cluster, Rancher will need fresh credentials ?

Can't we just update the credentials when it is needed for operations to take place ?

Thanks for your answers !

**Comment by dvyas1** [comment]
Ability to use EC2 instance profile credentials would be a very useful feature in small rancher deployments. It would also be extremely useful in dev/test environments. Its too bad this feature is not getting any priority.

**Comment by asemen2608** [comment]
Could we provide a fix for this problem? 
Customer hast this problem, when adding AWS cloud credentials that use SAML authentication the below error is displayed:

401 AuthFailure: There was a problem validating your keys. Enter valid credentials and try again.

**Comment by stacylacy6** [comment]
This continues to be a problem for our deployments as well.  We need a better solution than storing keys in the management plane.  Role based authentication is the primary method we use in other areas.  At a minimum, there needs to be a way to programatically update the keys so we can address key rotation in both the S3 backup and AWS credentials areas.  Managing around 30 clusters with Rancher, and now AWS key rotation for the S3 backups is becoming a heavy burden.
